File: shared/serialization/src/commonTest/kotlin/kmp/benchmarks/serialization/kotlinxserialization/KotlinxSerializationDecodeTest.kt
package kmp.benchmarks.serialization.kotlinxserialization

import kmp.benchmarks.serialization.JsonStrings
import kmp.benchmarks.serialization.model.kotlinxserialization.Coordinates
import kmp.benchmarks.serialization.model.kotlinxserialization.GeoJSONObject
import kmp.benchmarks.serialization.model.kotlinxserialization.HtmlChunk
import kmp.benchmarks.serialization.model.kotlinxserialization.MacOsReleases
import kotlinx.serialization.SerializationException
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class KotlinxSerializationDecodeTest {
    @Test
    fun decodeLargeListFromString() {
        KotlinxSerialization.decodeLargeListFromString("[]")

        val data = KotlinxSerialization.decodeLargeListFromString(JsonStrings.Minimised.largeList)

        assertEquals("2489651045", data.first().id)
        assertEquals("https://avatars.githubusercontent.com/u/665991?", data.first().actor.avatarUrl)
    }

    @Test
    fun decodeLargeListFromString_failed() {
        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeLargeListFromString("{}")
        }

        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeLargeListFromString("")
        }
    }

    @Test
    fun decodeLargeListCompactFromString() {
        KotlinxSerialization.decodeLargeListCompactFromString("[]")

        val data = KotlinxSerialization.decodeLargeListCompactFromString(JsonStrings.Minimised.largeList)

        assertEquals("2489651045", data.first().id)
        assertEquals("https://avatars.githubusercontent.com/u/665991?", data.first().actor.avatarUrl)
    }

    @Test
    fun decodeLargeListCompactFromString_failed() {
        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeLargeListCompactFromString("{}")
        }

        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeLargeListCompactFromString("")
        }
    }

    @Test
    fun decodeMacOsReleasesFromString() {
        val data = KotlinxSerialization.decodeMacOsReleasesFromString(JsonStrings.Minimised.macosReleases)

        assertEquals(MacOsReleases.Version.External.KODIAK, data.versions.first().external)
        assertEquals("US", data.versions.first().releases.first().version)
    }

    @Test
    fun decodeMacOsReleasesFromString_failed() {
        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeMacOsReleasesFromString("{}")
        }

        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeMacOsReleasesFromString("")
        }
    }

    @Test
    fun decodePolymorphicGeoFromString() {
        KotlinxSerialization.decodePolymorphicGeoFromString("[]")

        val data = KotlinxSerialization.decodePolymorphicGeoFromString(JsonStrings.Minimised.polymorphicGeo)

        assertTrue(data.first() is GeoJSONObject.Point)
        assertEquals(Coordinates(100.0, 0.0), (data.first() as GeoJSONObject.Point).coordinates)
    }

    @Test
    fun decodePolymorphicGeoFromString_failed() {
        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodePolymorphicGeoFromString("{}")
        }

        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodePolymorphicGeoFromString("")
        }
    }

    @Test
    fun decodePolymorphicHtmlFromString() {
        KotlinxSerialization.decodePolymorphicHtmlFromString("[]")

        val data = KotlinxSerialization.decodePolymorphicHtmlFromString(JsonStrings.Minimised.polymorphicHtml)

        assertTrue(data.first() is HtmlChunk.Paragraph)
        assertTrue((data.first() as HtmlChunk.Paragraph).data.text.isNotBlank())
    }

    @Test
    fun decodePolymorphicHtmlFromString_failed() {
        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodePolymorphicHtmlFromString("{}")
        }

        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodePolymorphicHtmlFromString("")
        }
    }

    @Test
    fun decodeUserProfileFromString() {
        val data = KotlinxSerialization.decodeUserProfileFromString(JsonStrings.Minimised.userProfile)

        assertTrue(data.personal.firstName.isNotBlank())
    }

    @Test
    fun decodeUserProfileFromString_failed() {
        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeUserProfileFromString("{}")
        }

        assertFailsWith<SerializationException> {
            KotlinxSerialization.decodeUserProfileFromString("")
        }
    }
}

File: app-ios/multiplatformTests/serialization/KotlinxSerializationDecodeTest.swift
//
//  KotlinxSerializationDecodeTest.swift
//  multiplatformTests
//
//  Created by Sebastian Owodzin on 12/01/2024.
//  Copyright © 2024 kmp. All rights reserved.
//

import BenchmarksShared
import XCTest

final class KotlinxSerializationDecodeTest: XCTestCase {
    func testDecodeLargeListFromString() throws {
        try! KotlinxSerialization().decodeLargeListFromString(jsonString: "[]")
        
        let data = try! KotlinxSerialization().decodeLargeListFromString(jsonString: JsonStrings.Minimised.largeList)
        
        XCTAssertEqual("2489651045", data.first!.id)
        XCTAssertEqual("https://avatars.githubusercontent.com/u/665991?", data.first!.actor.avatarUrl)
    }
    
    func testDecodeLargeListFromString_failed() throws {
        XCTAssertThrowsError(try KotlinxSerialization().decodeLargeListFromString(jsonString: "{}")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
        
        XCTAssertThrowsError(try KotlinxSerialization().decodeLargeListFromString(jsonString: "")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
    }
    
    func testDecodeLargeListCompactFromString() throws {
        try! KotlinxSerialization().decodeLargeListCompactFromString(jsonString: "[]")
        
        let data = try! KotlinxSerialization().decodeLargeListCompactFromString(jsonString: JsonStrings.Minimised.largeList)
        
        XCTAssertEqual("2489651045", data.first!.id)
        XCTAssertEqual("https://avatars.githubusercontent.com/u/665991?", data.first!.actor.avatarUrl)
    }
    
    func testDecodeLargeListCompactFromString_failed() throws {
        XCTAssertThrowsError(try KotlinxSerialization().decodeLargeListCompactFromString(jsonString: "{}")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
        
        XCTAssertThrowsError(try KotlinxSerialization().decodeLargeListCompactFromString(jsonString: "")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
    }

    func testDecodeMacOsReleasesFromString() throws {
        let data = try! KotlinxSerialization().decodeMacOsReleasesFromString(jsonString: JsonStrings.Minimised.macosReleases)

        XCTAssertEqual(MacOsReleases.VersionExternal.kodiak, data.versions.first!.external)
        XCTAssertEqual("US", data.versions.first!.releases.first!.version)
    }
    
    func testDecodeMacOsReleasesFromString_failed() throws {
        XCTAssertThrowsError(try KotlinxSerialization().decodeMacOsReleasesFromString(jsonString: "{}")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
        
        XCTAssertThrowsError(try KotlinxSerialization().decodeMacOsReleasesFromString(jsonString: "")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
    }

    func testDecodePolymorphicGeoFromString() throws {
        try! KotlinxSerialization().decodePolymorphicGeoFromString(jsonString: "[]")
        
        let data = try! KotlinxSerialization().decodePolymorphicGeoFromString(jsonString: JsonStrings.Minimised.polymorphicGeo)
        
        XCTAssertTrue(data.first is GeoJSONObjectPoint)
        XCTAssertEqual(Coordinates(longitude: 100.0, latitude: 0.0), (data.first! as! GeoJSONObjectPoint).coordinates)
    }
    
    func testDecodePolymorphicGeoFromString_failed() throws {
        XCTAssertThrowsError(try KotlinxSerialization().decodePolymorphicGeoFromString(jsonString: "{}")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
        
        XCTAssertThrowsError(try KotlinxSerialization().decodePolymorphicGeoFromString(jsonString: "")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
    }

    func testDecodePolymorphicHtmlFromString() throws {
        try! KotlinxSerialization().decodePolymorphicHtmlFromString(jsonString: "[]")
        
        let data = try! KotlinxSerialization().decodePolymorphicHtmlFromString(jsonString: JsonStrings.Minimised.polymorphicHtml)
        
        XCTAssertTrue(data.first is HtmlChunkParagraph)
        XCTAssertTrue((data.first as! HtmlChunkParagraph).data.text.isNotBlank())
    }
    
    func testDecodePolymorphicHtmlFromString_failed() throws {
        XCTAssertThrowsError(try KotlinxSerialization().decodePolymorphicHtmlFromString(jsonString: "{}")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
        
        XCTAssertThrowsError(try KotlinxSerialization().decodePolymorphicHtmlFromString(jsonString: "")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
    }
    
    func testDecodeUserProfileFromString() throws {
        let data = try! KotlinxSerialization().decodeUserProfileFromString(jsonString: JsonStrings.Minimised.userProfile)
        
        XCTAssertTrue(data.personal.firstName.isNotBlank())
    }
    
    func testDecodeUserProfileFromString_failed() throws {
        XCTAssertThrowsError(try KotlinxSerialization().decodeUserProfileFromString(jsonString: "{}")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
        
        XCTAssertThrowsError(try KotlinxSerialization().decodeUserProfileFromString(jsonString: "")) { error in
//            XCTAssertTrue(error is Kotlinx_serialization_coreSerializationException)
        }
    }
}
